### ISSUE TITLE
replace/rename support for DFS file path

### ISSUE THREAD

**Issue by mrmaxi** [jborean93/smbprotocol#85]
Thanks a lot for so great work!
I try to use your library with DFS file shares, but get some trouble.
"open_file", "listdir", "walk" and "remove" work correctly, but "rename" and "replace" cause STATUS_ACCESS_DENIED exception.
```
>>>smbclient.ClientConfig(username='user', password=’xxxx')

Traceback (most recent call last):
  File "C:/APPS/auto_env38/smbprotocol/t1.py", line 8, in <module>
    smbclient.replace('\\\\my-company.com\\DFS\\Report$\\Used\\Test\\Used\\FRB\\4.txt', '\\\\ my-company.com\\DFS\\Report$\\Used\\Test\\Used\\FRB\\6.txt')
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_os.py", line 488, in replace
    _rename_information(src, dst, replace_if_exists=True, **kwargs)
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_os.py", line 1047, in _rename_information
    set_info(transaction, file_rename)
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_io.py", line 237, in __exit__
    self.commit()
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_io.py", line 277, in commit
    raise failures[0]
smbprotocol.exceptions.SMBOSError: [Error 0] [NtStatus 0xc0000022] Unknown NtStatus error returned 'STATUS_ACCESS_DENIED': '\\my-company.com\DFS\Report$\Used\Test\Used\FRB\4.txt'
```

If I open_file().read() first, and then replace, it cause another exception “smbprotocol.exceptions.SMBOSError: [Error 2] [NtStatus 0xc000003a] No such file or directory”
```
print(smbclient.open_file('\\\\my-company.com\\DFS\\Report$\\Used\\Test\\Used\\FRB\\4.txt').read())
smbclient.replace('\\\\my-company.com\\DFS\\Report$\\Used\\Test\\Used\\FRB\\4.txt', '\\\\my-company.com\\DFS\\Report$\\Used\\Test\\Used\\FRB\\6.txt')

Hello world
Traceback (most recent call last):
  File "C:/APPS/auto_env38/smbprotocol/t1.py", line 8, in <module>
    smbclient.replace('\\\\my-company.com\\DFS\\Report$\\Used\\Test\\Used\\FRB\\4.txt', '\\\\my-company.com\\DFS\\Report$\\Used\\Test\\Used\\FRB\\6.txt')
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_os.py", line 488, in replace
    _rename_information(src, dst, replace_if_exists=True, **kwargs)
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_os.py", line 1047, in _rename_information
    set_info(transaction, file_rename)
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_io.py", line 237, in __exit__
    self.commit()
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_io.py", line 277, in commit
    raise failures[0]
smbprotocol.exceptions.SMBOSError: [Error 2] [NtStatus 0xc000003a] No such file or directory: '\\my-company.com\DFS\Report$\Used\Test\Used\FRB\4.txt'
```

If I execute replace with direct path (instead of using DFS), it work perfectly:
`smbclient.replace('\\\\fs-p24\\Report$\\Used\\Test\\Used\\FRB\\5.txt', '\\\\fs-p24\\Report$\\Used\\Test\\Used\\FRB\\6.txt')`

I understand that DFS support is experimental, but is replace through DFS not implemented, or just I made something wrong?

**Comment by mrmaxi** [comment]
I try on Windows and Linux both, 1.4 and 1.5 smbprotocol version

```
>>>python
Python 3.8.7 (tags/v3.8.7:6503f05, Dec 21 2020, 17:59:51) [MSC v.1928 64 bit (AMD64)] on win32

>>>pip list
Package      Version
------------ -------
appdirs      1.4.4
atomicwrites 1.4.0
attrs        20.3.0
cffi         1.14.5
colorama     0.4.4
coverage     5.5
cryptography 3.4.6
distlib      0.3.1
filelock     3.0.12
iniconfig    1.1.1
packaging    20.9
pip          21.0.1
pluggy       0.13.1
py           1.10.0
pycodestyle  2.7.0
pycparser    2.20
pyparsing    2.4.7
pyspnego     0.1.5
pytest       6.2.2
pytest-cov   2.11.1
pytest-mock  3.5.1
setuptools   54.1.2
six          1.15.0
smbprotocol  1.5.0
toml         0.10.2
tox          3.23.0
virtualenv   20.4.3
wheel        0.36.2
```

**Comment by mrmaxi** [comment]
I found out that remove and open file for write by DFS file path working only if I open file for read from the same DFS root earlier.
for example:
```
smbclient.ClientConfig(username='user', password='xxxx'))
print(smbclient.open_file('\\\\my-company.com\\DFS\\Report$\\Used\\ASV\\tmp\\CTR.TXT').read())
print(smbclient.open_file('\\\\my-company.com\\DFS\\Report$\\Used\\Test\\Used\\FRB\\5.txt', 'w').write('55111'))
```
work correctly, but
```
smbclient.ClientConfig(username='user', password='xxxx'))
# print(smbclient.open_file('\\\\my-company.com\\DFS\\Report$\\Used\\ASV\\tmp\\CTR.TXT').read())
print(smbclient.open_file('\\\\my-company.com\\DFS\\Report$\\Used\\Test\\Used\\FRB\\5.txt', 'w').write('55111'))
```
fails with exception:
```
Traceback (most recent call last):
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_io.py", line 368, in open
    open_result = self.fd.create(
  File "C:\APPS\auto_env38\smbprotocol\smbprotocol\open.py", line 1213, in create
    return self._create_response(request)
  File "C:\APPS\auto_env38\smbprotocol\smbprotocol\open.py", line 1219, in _create_response
    response = self.connection.receive(request)
  File "C:\APPS\auto_env38\smbprotocol\smbprotocol\connection.py", line 928, in receive
    raise SMBResponseException(response)
smbprotocol.exceptions.AccessDenied: Received unexpected status from the server: A process has requested access to an object but has not been granted those access rights. (3221225506) STATUS_ACCESS_DENIED: 0xc0000022

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:/APPS/auto_env38/smbprotocol/t1.py", line 9, in <module>
    print(smbclient.open_file('\\\\bank-hlynov.ru\\DFS\\Report$\\Used\\Test\\Used\\FRB\\5.txt', 'w').write('55111'))
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_os.py", line 365, in open_file
    raw_fd.open()
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_io.py", line 412, in open
    raise SMBOSError(exc.status, self.name)
smbprotocol.exceptions.SMBOSError: [Error 0] [NtStatus 0xc0000022] Unknown NtStatus error returned 'STATUS_ACCESS_DENIED': '\\my-company.com\DFS\Report$\Used\Test\Used\FRB\5.txt'
```

**Comment by jborean93** [comment]
Thanks for the bug report, I'll have to set up a DFS endpoint to test this out. For some background information the replace/rename operation involves 2 steps

* Opens the source file that is to be replaced/renamed
* Sends a [FileRenameInformation](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-fscc/52aa0b70-8094-4971-862d-79793f41e6a8) request on this handle with `FileName` being the value of `dest`.

In this case it looks like it's the 2nd step that is failing so something is up there. I'll really have to try to replicate it myself to be able to understand what may be happening. What would be good to know is if you can replace/rename the file using a native Windows command. I assume it is fine but if you are able to try this out on a Windows host that can access your DFS path it would be great to get confirmation.

```powershell
[IO.File]::Move(
    '\\my-company.com\DFS\Report$\Used\Test\Used\FRB\4.txt',
    '\\my-company.com\DFS\Report$\Used\Test\Used\FRB\6.txt'
)
```

**Comment by mrmaxi** [comment]
Thank you for fas reply!
I execute command in power shell, all work fine.
```
(base) PS C:\Users\mrmaxi> [IO.File]::Move(
'\\my-company.com\DFS\Report$\Used\Test\Used\FRB\4.txt', 
'\\my-company.com\DFS\Report$\Used\Test\Used\FRB\6.txt')
(base) PS C:\Users\mrmaxi>
```
File was renamed correctly.

**Comment by jborean93** [comment]
Cool I think I've found the problem(s) and I'm trying to think of a way to solve them without adding too much complexity. The problems I see are:

* The internal transaction handler (what sends multiple SMB requests as 1) does not handle `STATUS_PATH_NOT_COVERED` for DFS paths
* The logic to build the `FileRenameInformation` could be better when it comes to setting the replacement path for DFS paths

The first problem helps to explain why you are seeing different behaviour if you read the file beforehand. By reading the file you are caching the DFS referrals used to get that path which means the transaction open works properly.

**Comment by jborean93** [comment]
I've created a PR which I believe fixes this problem for you https://github.com/jborean93/smbprotocol/pull/86. It would be great if you could test it with your environment to see if it fixes your problem.

**Comment by mrmaxi** [comment]
Thanks a lot! 
With #86 rename/replace work fine in my environment.

Trouble with need to open file for read before try to write/remove/rename/replace it still exists.
But I already can use the library with this workaround, thanks!

**Comment by mrmaxi** [comment]
Sorry, I hurried to answer. I still has STATUS_ACCESS_DENIED when I try to replace file.
```python
 smbclient.replace('\\\\my-company.com\\DFS\\Report$\\Used\\Test\\Used\\FRB\\arch\\yar01.DBF','\\\\my-company.com\\DFS\\Report$\\Used\\Test\\Used\\FRB\\arch\\tmp\\yar01.DBF')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.8/site-packages/smbprotocol-1.5.1-py3.8.egg/smbclient/_os.py", line 488, in replace
  File "/usr/local/lib/python3.8/site-packages/smbprotocol-1.5.1-py3.8.egg/smbclient/_os.py", line 1051, in _rename_information
  File "/usr/local/lib/python3.8/site-packages/smbprotocol-1.5.1-py3.8.egg/smbclient/_io.py", line 410, in __enter__
  File "/usr/local/lib/python3.8/site-packages/smbprotocol-1.5.1-py3.8.egg/smbclient/_io.py", line 465, in open
  File "/usr/local/lib/python3.8/site-packages/smbprotocol-1.5.1-py3.8.egg/smbclient/_io.py", line 359, in commit
smbprotocol.exceptions.SMBOSError: [Error 0] [NtStatus 0xc0000022] Unknown NtStatus error returned 'STATUS_ACCESS_DENIED': '\\my-company.com\DFS\Report$\Used\Test\Used\FRB\arch\yar01.DBF'
```
Sometimes repeat "replace" command work fine, sometimes I retry many times and everytime has STATUS_ACCESS_DENIED

**Comment by jborean93** [comment]
The only reason why that would return with `STATUS_ACCESS_DENIED` is if the SMB user doesn't have `DELETE` rights to the file. This occurs before the replace operation occurs and it's just when the file is opened. For example `smbclient` is opening the file with just the `DELETE` rights as specified at https://github.com/jborean93/smbprotocol/blob/dfs-replace/smbclient/_os.py#L1039.

If the file was already opened by another process without the `Delete` file share set by that process then this operation should fail with `STATUS_SHARING_VIOLATION`.

You can hopefully replicate this call by running this PowerShell snippet

```powershell
Add-Type -TypeDefinition @'
using Microsoft.Win32.SafeHandles;
using System;
using System.ComponentModel;
using System.IO;
using System.Runtime.InteropServices;

namespace Win32
{
    public class NativeMethods
    {
        [DllImport("Kernel32.dll", CharSet = CharSet.Unicode, EntryPoint = "CreateFileW", SetLastError = true)]
        private static extern SafeFileHandle CreateFileRaw(
            string lpFileName,
            int dwDesiredAccess,
            FileShare dwShareMode,
            IntPtr lpSecurityAttributes,
            FileMode dwCreationDisposition,
            int dwFlagsAndAttributes,
            IntPtr hTemplateFile);
            
        public static SafeFileHandle CreateFile(string lpFileName, int desiredAccess, FileShare share,
            FileMode mode, int flagsAndAttributes)
        {
            SafeFileHandle handle = CreateFileRaw(lpFileName, desiredAccess, share, IntPtr.Zero, mode,
                flagsAndAttributes, IntPtr.Zero);
                
            if (handle.IsInvalid)
                throw new Win32Exception(Marshal.GetLastWin32Error());
                
            return handle;
        }
    }
}
'@

$handle = [Win32.NativeMethods]::CreateFile(
    '\\my-company.com\DFS\Report$\Used\Test\Used\FRB\arch\yar01.DBF',
    0x10000,  # DELETE
    'ReadWrite, Delete',
    'Open',
    0x00200000  # FILE_FLAG_OPEN_REPARSE_POINT
)
$handle.Dispose()
```

**Comment by mrmaxi** [comment]
If i right undestand for find out what happens realy I does:
1. execute powershell (**I remove `Delete` from share argument of your example**):
```PowerShell
$handle = [Win32.NativeMethods]::CreateFile(
    '\\my-company.com\DFS\Report$\Used\Test\Used\FRB\arch\yar01.DBF',
    0x10000,  # DELETE
    'ReadWrite',
    'Open',
    0x00200000  # FILE_FLAG_OPEN_REPARSE_POINT
)
```
2. try to rename file `yar01.DBF` by windows explorer. As expected I get an exception about file is being used by another process.
3. try to rename by smbclient. I get an exception as earlier `STATUS_ACCESS_DENIED`
```python
smbclient.replace(r'\\my-company.com\DFS\Report$\Used\Test\Used\FRB\arch\yar01.DBF', r'\\my-company.com\DFS\Report$\Used\Test\Used\FRB\arch\yar02.DBF') 
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_os.py", line 488, in replace
    _rename_information(src, dst, replace_if_exists=True, **kwargs)
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_os.py", line 1051, in _rename_information
    with src_raw:
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_io.py", line 410, in __enter__
    self.open()
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_io.py", line 465, in open
    transaction.commit()
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_io.py", line 359, in commit
    raise failures[0]
smbprotocol.exceptions.SMBOSError: [Error 0] [NtStatus 0xc0000022] Unknown NtStatus error returned 'STATUS_ACCESS_DENIED': '\\my-company.com\DFS\Report$\Used\Test\Used\FRB\arch\yar01.DBF'
```
4. I retry attempt, and get another exception:
```python
smbclient.replace(r'\\my-company.com\DFS\Report$\Used\Test\Used\FRB\arch\yar01.DBF', r'\\my-company.com\DFS\Report$\Used\Test\Used\FRB\arch\yar02.DBF') 
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_os.py", line 488, in replace
    _rename_information(src, dst, replace_if_exists=True, **kwargs)
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_os.py", line 1051, in _rename_information
    with src_raw:
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_io.py", line 410, in __enter__
    self.open()
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_io.py", line 465, in open
    transaction.commit()
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_io.py", line 359, in commit
    raise failures[0]
smbprotocol.exceptions.SMBOSError: [Error 1] [NtStatus 0xc0000043] The process cannot access the file because it is being used by another process: '\\my-company.com\DFS\Report$\Used\Test\Used\FRB\arch\yar01.DBF'
```
5. Then I dispose file handle:
```PowerShell
$handle.Dispose()
```
6. Retry by smbclient and **all work fine**:
```python
smbclient.replace(r'\\my-company.com\DFS\Report$\Used\Test\Used\FRB\arch\yar01.DBF', r'\\my-company.com\DFS\Report$\Used\Test\Used\FRB\arch\yar02.DBF')
```
7. wait couple of minutes
8. try to rename backward yar02.DBF -> yar01.DBF:
```python
smbclient.replace(r'\\my-company.com\DFS\Report$\Used\Test\Used\FRB\arch\yar02.DBF', r'\\my-company.com\DFS\Report$\Used\Test\Used\FRB\arch\yar01.DBF') 
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_os.py", line 488, in replace
    _rename_information(src, dst, replace_if_exists=True, **kwargs)
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_os.py", line 1051, in _rename_information
    with src_raw:
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_io.py", line 410, in __enter__
    self.open()
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_io.py", line 465, in open
    transaction.commit()
  File "C:\APPS\auto_env38\smbprotocol\smbclient\_io.py", line 359, in commit
    raise failures[0]
smbprotocol.exceptions.SMBOSError: [Error 0] [NtStatus 0xc0000022] Unknown NtStatus error returned 'STATUS_ACCESS_DENIED': '\\my-company.com\DFS\Report$\Used\Test\Used\FRB\arch\yar02.DBF'
```
9. I retry several times - the error persists. `STATUS_ACCESS_DENIED` everytime
10. I try `smbclient.reset_connection_cache()`, but nothing changes, when I try rename - `STATUS_ACCESS_DENIED` occured
11. If I re-run python and execute replace again, I as earlier at first time have `STATUS_ACCESS_DENIED`, but at second time - **all works fine**

**Comment by jborean93** [comment]
Ok at this point I think we need a network capture of the `smbclient` exchange to see what exactly is causing the `STATUS_ACCESS_DENIED` in relation to the DFS work. Are you able to run `tcpdump -i <interface> -nn port 445 -X -q -w capture.pcap` where `<interface>` is the network interface name, like `enp5s0`, that the SMB traffic is run over. You can also use a tool like `Wireshark` to capture this information if that's easier for you.

This can contain sensitive information so feel free to email the details, if you feel comfortable, to me. Hopefully this can illuminate what the exact request was that causes the `STATUS_ACCESS_DENIED` and then the one that causes `STATUS_SHARING_VIOLATION`.

**Comment by jborean93** [comment]
So I think the problem you have here is also described in https://github.com/hierynomus/smbj/issues/328, what I do not understand is how to replicate the problem myself. No matter how hard I try I cannot get a `STATUS_ACCESS_DENIED` error when trying to open a file on a DFS share and I cannot find anything in the DFS specifications around this particular error.

In saying that I've updated https://github.com/jborean93/smbprotocol/pull/86 to hopefully stop this from occurring in `replace/rename`. It would be great if you could try the changes there again and see if it solves the `STATUS_ACCESS_DENIED` error you are getting.

**Comment by mrmaxi** [comment]
Hello Jordan!
I don't understand why, but with last commit in #86 replace/rename work! 
However, I still get `STATUS_ACCESS_DENIED` when run replace/rename in the same session two minutes later .

For example:
1. replace file1->file2. all right
2. replace file2->file1. all right
3. sleep 61 seconds
4. replace file1->file2. `STATUS_ACCESS_DENIED`

It's looks like some kind of timeout, because if I `sleep(59)` at step 3 - `replace` work fine.

**Comment by jborean93** [comment]
When looking at the packet capture you sent through I can see the time to live (TIL) is set to 60 seconds so it makes sense that 60 second timeout stops it from working again. The referral is invalidated and thus it tries to connect over the original path. What I just don't understand is why it isn't returning `STATUS_PATH_NOT_COVERED` to let the client know that the file is on a DFS share and you it needs to send the referral.

* Is there anything special about this file, is it hidden, a symlink, read only, or anything else you could see as special
* What the DFS server you are connecting to, is a Windows host, a Samba host, Netapp, or some other service?
* Does it help in any way if you set `smbclient.ClientConfig(domain_controller='my-dc.realm.com')` to precache the domain DFS servers
* Can you run the following and see if it ever fails with `STATUS_ACCESS_DENIED`

```python
import smbclient

# Set this to your source path
file_path = r'\\domain\dfs\folder\file.txt'
with smbclient.open_file(file_path, mode='rb', desired_access=0x00010000, create_options=0x00200000) as fd:
    pass
```

**Comment by mrmaxi** [comment]
- file is ordinary, nothing special. I create it in notepad.
- **with `smbclient.ClientConfig(domain_controller='my-dc.realm.com')` work fine!** 
I confused, because earely I read about this in #59, and try with domain_controller, but at that time it does not help.

Jordan, thank you. 
I have one trouble more with this `STATUS_ACCESS_DENIED`. If I open_file on DFS for **write** at once, I has `STATUS_ACCESS_DENIED`. But if I open_file for read at first, and than for write, all work fine. I try with `domain_controller` and without both.
I in doubt, is this the same trouble, or I have to open another issue?

**Comment by jborean93** [comment]
Unfortunately I cannot fathom what the problem is there. I'll try out a few more things but I'm running out of things to try out.

So I think https://github.com/jborean93/smbprotocol/pull/86 fixes an actual problem but ultimately your problem seems to be something else. For whatever reason trying to access a file on DFS host itself is returning `STATUS_ACCESS_DENIED` rather than `STATUS_PATH_NOT_COVERED`. The latter is what is used to determine whether we need to resolve the DFS path or not. Anything else is not part of the spec and I don't think we can really use `STATUS_ACCESS_DENIED` to indicate this as well.

Do you know what is hosting your DFS share. Is it a Windows Server, NetApp, Samba, or some other implementation. Maybe the reason I haven't been able to replicate it is because it's not something that Windows does and I need to try out that other product.

**Comment by mrmaxi** [comment]
I forgot to write, my DFS host is Windows Server 2012 R2. Admin of this DFS-host and DC is colleague of mine, so I can check something if it needed.

**Comment by mrmaxi** [comment]
I made some experiments.
```python
with smbclient.open_file('\\my-company.com\DFS\test\1.txt', mode='rb', desired_access=0x00010000, create_options=0x00200000) as fd:
    pass
```
fails by `STATUS_ACCESS_DENIED`
At the same time without `desired_access=0x00010000` - work fine
I compare wireshark traffic, it seems like only one difference is:
```
Ioctl Request FSCTL_VALIDATE_NEGOTIATE_INFO
Ioctl Response FSCTL_VALIDATE_NEGOTIATE_INFO
Create Request File: test\1.txt
  Access Mask: 0x00000089
```
response:
```
Create Response, Error: STATUS_PATH_NOT_COVERED
```
But in second case:
```
Ioctl Request FSCTL_VALIDATE_NEGOTIATE_INFO
Ioctl Response FSCTL_VALIDATE_NEGOTIATE_INFO
Create Request File: test\1.txt
Access Mask: 0x00010000
```
response:
```
Create Response, Error: STATUS_ACCESS_DENIED
```

It looks strange, so I try, dump traffic from request by PowerShell
```PowerShell
  $handle = [Win32.NativeMethods]::CreateFile(
     '\\my-company.com\DFS\test\1.txt',
     0x10000,  # DELETE
     'Read',
     'Open',
     0x00200000  # FILE_FLAG_OPEN_REPARSE_POINT
)
```
I find out that PowerShell does not execute `Create Request File` at once. PowerShell made `FSCTL_DFS_GET_REFERRALS` before `Create Request File`, without waiting for `STATUS_PATH_NOT_COVERED`.
```
Ioctl Request FSCTL_VALIDATE_NEGOTIATE_INFO
Ioctl Response FSCTL_VALIDATE_NEGOTIATE_INFO
Ioctl Request FSCTL_QUERY_NETWORK_INTERFACE_INFO
Ioctl Request FSCTL_DFS_GET_REFERRALS, File: \FS-U40\dfs\test
Ioctl Response FSCTL_QUERY_NETWORK_INTERFACE_INFO
Create Request File: 1.txt
Create Response File: 1.txt
Close Request File: 1.txt
Close Response
```
Is there another way for find out that it's need to get dfs refferals except of receive `STATUS_PATH_NOT_COVERED`?

**Comment by jborean93** [comment]
Thanks for the further info, definitely helps.

> Is there another way for find out that it's need to get dfs refferals except of receive STATUS_PATH_NOT_COVERED?

That’s what `smbclient.ClientConfig(domain_controller='dc')` is meant to help with. It will lookup the domain controller to find the DFS namespaces so when it goes to connect to `\\domain\root\...` it knows it’s a DFS path and it to send the referral first before connecting. Without that smbclient just thinks you are trying to connect to the host called `domain` which is most likely the DC and then relies on the `STATUS_PATH_NOT_COVERED` result to send the DFS referral if received. In this case it thinks it’s a normal path and because it received `STATUS_ACCESS_DENIED` it just thinks you don’t have the necessary rights for that path.

The Windows SMB client is a whole lot more complex and it essentially runs as a service so the cache persists beyond the process where you connected to it. It runs many things in the background and by default the Windows hosts know if you are connected to a domain so it can automatically talk to the DC to get the DFS namespaces automatically.

This is why I’m so perplexed and cannot replicate the problem on my end because the DFS docs do not mention the access error at all when it comes to resolving a path. The only place where this is mentioned is the Java smbj library and even then it’s a bit vague.

**Comment by jborean93** [comment]
I've merged https://github.com/jborean93/smbprotocol/pull/86 and while it doesn't completely solve this issue it does fix up a few problems that were brought up in this issue.

I've tried in vain to replicate this behaviour myself and unfortunately I have been unable to do so. If you do come across any more information that might help to understand the root problem I'm happy to hear it but for now there's not much else I can do so I'm closing the issue.
